File: src/database/supabase/storage/reviewStorage.js
import { supabase, BUCKETS } from '../config';

/**
 * UPLOAD REVIEW IMAGE
 */
export const uploadReviewImage = async (reviewId, file, imageIndex) => {
  try {
    const fileName = `${reviewId}/${imageIndex}.jpg`;
    
    const { data, error } = await supabase.storage
      .from(BUCKETS.REVIEW_IMAGES)
      .upload(fileName, file, {
        cacheControl: '3600',
        upsert: true
      });

    if (error) throw error;

    const { data: publicUrlData } = supabase.storage
      .from(BUCKETS.REVIEW_IMAGES)
      .getPublicUrl(fileName);

    return { 
      success: true, 
      url: publicUrlData.publicUrl 
    };
  } catch (error) {
    return { 
      success: false, 
      error: error.message 
    };
  }
};

/**
 * UPLOAD MULTIPLE REVIEW IMAGES
 */
export const uploadMultipleReviewImages = async (reviewId, files) => {
  try {
    const uploadPromises = files.map((file, index) => 
      uploadReviewImage(reviewId, file, index)
    );
    
    const results = await Promise.all(uploadPromises);
    
    const failedUploads = results.filter(r => !r.success);
    if (failedUploads.length > 0) {
      throw new Error('Some images failed to upload');
    }
    
    const urls = results.map(r => r.url);
    
    return { 
      success: true, 
      urls 
    };
  } catch (error) {
    return { 
      success: false, 
      error: error.message 
    };
  }
};

/**
 * DELETE REVIEW IMAGE
 */
export const deleteReviewImage = async (reviewId, imageIndex) => {
  try {
    const fileName = `${reviewId}/${imageIndex}.jpg`;
    
    const { error } = await supabase.storage
      .from(BUCKETS.REVIEW_IMAGES)
      .remove([fileName]);

    if (error) throw error;

    return { success: true };
  } catch (error) {
    return { 
      success: false, 
      error: error.message 
    };
  }
};

/**
 * DELETE ALL REVIEW IMAGES
 */
export const deleteAllReviewImages = async (reviewId, imageCount) => {
  try {
    const fileNames = Array.from({ length: imageCount }, (_, i) => 
      `${reviewId}/${i}.jpg`
    );
    
    const { error } = await supabase.storage
      .from(BUCKETS.REVIEW_IMAGES)
      .remove(fileNames);

    if (error) throw error;

    return { success: true };
  } catch (error) {
    return { 
      success: false, 
      error: error.message 
    };
  }
};

File: app/src/main/java/com/example/bunnix/database/supabase/storage/ReviewStorage.kt
package com.example.bunnix.database.supabase.storage

import android.content.Context
import android.net.Uri
import com.example.bunnix.database.config.SupabaseConfig
import io.github.jan.supabase.storage.storage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File

object ReviewStorage {

    private val storage = SupabaseConfig.client.storage

    // UPLOAD REVIEW IMAGE
    suspend fun uploadReviewImage(
        context: Context,
        reviewId: String,
        imageUri: Uri,
        imageIndex: Int
    ): Result<String> = withContext(Dispatchers.IO) {
        try {
            val file = uriToFile(context, imageUri)
            val fileBytes = file.readBytes()
            val path = "$reviewId/$imageIndex.jpg"

            val bucket = storage.from(SupabaseConfig.Buckets.REVIEW_IMAGES)

            bucket.upload(
                path = path,
                data = fileBytes,
                options = {
                    upsert = true
                }
            )

            val publicUrl = bucket.publicUrl(path)
            file.delete()

            Result.success(publicUrl)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // UPLOAD MULTIPLE REVIEW IMAGES
    suspend fun uploadMultipleReviewImages(
        context: Context,
        reviewId: String,
        imageUris: List<Uri>
    ): Result<List<String>> = withContext(Dispatchers.IO) {
        try {
            val uploadedUrls = mutableListOf<String>()

            imageUris.forEachIndexed { index, uri ->
                val result = uploadReviewImage(context, reviewId, uri, index)
                result.onSuccess { url ->
                    uploadedUrls.add(url)
                }
                result.onFailure { error ->
                    throw error
                }
            }

            Result.success(uploadedUrls)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // DELETE REVIEW IMAGE
    suspend fun deleteReviewImage(reviewId: String, imageIndex: Int): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            val path = "$reviewId/$imageIndex.jpg"
            val bucket = storage.from(SupabaseConfig.Buckets.REVIEW_IMAGES)
            bucket.delete(path)
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // DELETE ALL REVIEW IMAGES
    suspend fun deleteAllReviewImages(reviewId: String, imageCount: Int): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            val paths = (0 until imageCount).map { "$reviewId/$it.jpg" }
            val bucket = storage.from(SupabaseConfig.Buckets.REVIEW_IMAGES)
            bucket.delete(paths)
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    private fun uriToFile(context: Context, uri: Uri): File {
        val inputStream = context.contentResolver.openInputStream(uri)
            ?: throw Exception("Cannot open URI")

        val tempFile = File(context.cacheDir, "temp_${System.currentTimeMillis()}.jpg")
        tempFile.outputStream().use { output ->
            inputStream.copyTo(output)
        }
        inputStream.close()

        return tempFile
    }
}